woocommerce/woocommerce-ios
# [Mobile Payments] Use selected gateway for In-Person Payments

Closes: #7086 


### Description


This makes the payment flow actually use the stored preference for payments when there is more than one extension active.

It also ensures that we disconnect any connected readers when we select a different plugin to use. Connection tokens are tied to a specific account, and won't work if we switch to a different one, so we need to ensure we request a new one from the right account.

On https://github.com/woocommerce/woocommerce-ios/pull/7087 @koke implemented a first approach to solve this issue,  including the selected plugin as part of the `CardPaymentReadiness` state, and using it to retrieve the `PaymentGatewayAccount` from storage when collecting a payment in the Order Details screen.
However, as he mentioned in that PR, this approach was not generic enough and wouldn't work optimally for the other two cases that we have, such as refunds.
With this PR we want to add a more generic approach to this issue. As we know, `CardPresentPaymentsOnboardingUseCase` always dispatches `CardPresentPaymentAction.use` with the selected `PaymentGatewayAccount` at the end of the account checks before returning a `complete` state, so `CardPresentPaymentStore` should always have an up-to-date selection. With this in mind, we refactor the `CardPresentPaymentStore` to keep a reference of that gateway account so it can be retrieved via a `CardPresentPaymentAction` that is also included here. In this manner, we can retrieve the select payment gateway account for every case we need it instead of passing the plugin info from the onboarding.
Furthermore, now that we have the selected `PaymentGatewayAccount` in `CardPresentPaymentStore` we can use it to determine which backend to use at any moment with `usingBackend`, that is now a computed property.
Also, as we now retrieve the selected `PaymentGatewayAccount`from the store, we can remove here the code used to retrieve the accounts from `CoreData` via `ResultsController`

### Testing instructions

Use a store that has both WCPay and Stripe extensions installed and set up.

1. Go to Settings > In-Person Payments
2. In Payment Options, make sure WCPay is selected
3. Go to orders, create a simple payment and pay with a card
4. When you go to the order details, it should say "Paid via WooCommerce In-Person Payments"
5. Go to Settings > In-Person Payments
6. In Payment Options, make sure Stripe is selected
7. Go to orders, create a simple payment and pay with a card. **Notice the reader has been disconnected and automatically reconnects.**
8. When you go to the order details, it should say "Paid via WooCommerce Stripe In-Person Payments"

Repeat steps for refunds.

### Screenshots


Paid with WCPay | Paid with Stripe
-|-
![Screen Shot 2022-06-15 at 16 42 40](https://user-images.githubusercontent.com/8739/173859615-4b3f54f4-c534-42f7-8d58-b383d0a00576.png) | ![Screen Shot 2022-06-15 at 16 56 29](https://user-images.githubusercontent.com/8739/173859637-7f3d7dcc-626a-4303-b559-f873b6bf0d76.png)

---
